src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestAttributeValueResolver.php

```diff
@@ -27,7 +27,13 @@
      */
     public function supports(Request $request, ArgumentMetadata $argument)
     {
-        return !$argument->isVariadic() && $request->attributes->has($argument->getName());
+        if ($argument->isVariadic()) {
+            return false;
+        }
+
+        $argumentName = $argument->getName();
+
+        return $request->attributes->has($argumentName) && null !== $request->attributes->get($argumentName);
```

Comment:
Using: 

```php
/**
 * @Route("/", name="homepage", defaults={"required": null})
 */
public function indexAction(string $required)
{
}
```

> Controller "AppBundle\Controller\DefaultController::indexAction()" requires that you provide a value for the "$required" argument [...]

Changing the signature to `fooAction(string $required = null);` or `fooAction(?string $required);` works of course.

Same behavior on 3.0, run into the same exception if the attribute is null.